File: Classes/THRuleEditorView/THValueAndUnitView.m
// THValueAndUnitView.m

#import "THValueAndUnitView.h"

//--------------------------------------------------------------------------------------------------------------------------------------------
@interface THValueAndUnitView_PopUpButton : NSPopUpButton
@end

@implementation THValueAndUnitView_PopUpButton
- (BOOL)canBecomeKeyView { return YES; }
@end

@interface THValueAndUnitView_Stepper : NSStepper
@end

@implementation THValueAndUnitView_Stepper
- (BOOL)canBecomeKeyView { return YES; }
@end
//--------------------------------------------------------------------------------------------------------------------------------------------


//--------------------------------------------------------------------------------------------------------------------------------------------
@implementation THValueAndUnitView

- (id)initWithFrame:(NSRect)frame popBezelStyle:(NSBezelStyle)popBezelStyle controlView:(id<THValueAndUnitViewDelegateProtocol>)controlView
{
	if (self=[super initWithFrame:frame])
	{
		_controlView=controlView;

		NSRect fieldRect=NSMakeRect(0.0,CGFloatFloor((frame.size.height-19.0)/2.0),50.0,19.0);
		
		NSRect stepperRect=NSMakeRect(fieldRect.size.width+4.0,CGFloatFloor((frame.size.height-22.0)/2.0),16.0,22.0);

		NSRect menuRect=NSMakeRect(stepperRect.origin.x+stepperRect.size.width+4.0,CGFloatFloor((frame.size.height-22.0)/2.0),60.0,22.0);
		if ([self isKindOfClass:[THDateWithinStepperView class]]==YES)
			menuRect.size.width=90.0;
	
		NSFont *font=[NSFont systemFontOfSize:[NSFont systemFontSizeForControlSize:NSControlSizeSmall]];

		{
			_textField=[[NSTextField alloc] initWithFrame:fieldRect];

			_textField.font=font;
			//	result.target=self;
			//	result.action=@selector(criterionTextFieldAction:);
			_textField.delegate=self;
			[_textField setEditable:YES];
			[_textField setSelectable:YES];
			[_textField setDrawsBackground:YES];

			NSNumberFormatter *nbFormatter=[[NSNumberFormatter alloc] init];
			nbFormatter.numberStyle=NSNumberFormatterDecimalStyle;
			nbFormatter.usesGroupingSeparator=NO;
			_textField.formatter=nbFormatter;

			[(NSTextFieldCell*)_textField.cell setControlSize:NSControlSizeSmall];
			[(NSTextFieldCell*)_textField.cell setAlignment:NSTextAlignmentRight];

			[self addSubview:_textField];
		}

		{
			NSDictionary *stepInfos=[self valueStepperInfos];
			_stepper=[[THValueAndUnitView_Stepper alloc] initWithFrame:stepperRect];
			_stepper.target=self;
			_stepper.action=@selector(fileSizeStepperAction:);
			_stepper.minValue=1.0;
			_stepper.integerValue=1;
			_stepper.maxValue=[stepInfos[@"maxValue"] integerValue];
			[(NSStepperCell*)_stepper.cell setControlSize:NSControlSizeSmall];
			
			[self addSubview:_stepper];
		}

		{
			NSMenu *menu=[self unitMenu];
			_popUpButton=[[THValueAndUnitView_PopUpButton alloc] initWithFrame:menuRect];
			_popUpButton.alignment=NSTextAlignmentLeft;
			_popUpButton.bezelStyle=popBezelStyle;
			_popUpButton.target=self;
			_popUpButton.action=@selector(fileSizePoidsMenuAction:);
			_popUpButton.menu=menu;
			_popUpButton.autoresizingMask=NSViewMaxXMargin;

			[(NSPopUpButtonCell*)_popUpButton.cell setControlSize:NSControlSizeSmall];
			[(NSPopUpButtonCell*)_popUpButton.cell setFont:font];

			[self addSubview:_popUpButton];
		}
	}
	return self;
}

- (void)dealloc
{
    _controlView=nil;
}

#pragma mark -

- (NSDictionary*)valueStepperInfos { return nil; }

- (NSMenu*)unitMenu { return nil; }

#pragma mark -

- (NSInteger)value { return _textField.integerValue; }

- (NSInteger)poids { return _popUpButton.selectedTag; }

- (void)setValue:(NSInteger)value poids:(NSInteger)poids
{
	_textField.integerValue=value;
	_stepper.integerValue=value;
	[_popUpButton selectItemWithTag:poids];
}

- (NSString*)valueComps
{
	return [NSString stringWithFormat:@"%d|%d",(int)[self value],(int)[self poids]];
}

- (void)setValueComps:(NSString*)comps
{
	NSArray *c=[comps componentsSeparatedByString:@"|"];
	[self setValue:c.count==2?[c[0] integerValue]:1 poids:c.count==2?[c[1] integerValue]:1];
}

- (void)controlTextDidChange:(NSNotification*)notification
{
	if (notification.object!=_textField)
		return;
	if (_textField.integerValue>=(NSInteger)_stepper.minValue)
		_stepper.integerValue=_textField.integerValue;
	[_controlView valueAndUnitViewDidChange:self];
}

- (void)fileSizeStepperAction:(NSStepper*)sender
{
	_textField.integerValue=sender.integerValue;
	[_controlView valueAndUnitViewDidChange:self];
}

- (void)fileSizePoidsMenuAction:(NSPopUpButton*)sender
{
	[_controlView valueAndUnitViewDidChange:self];
}

- (void)setEnabled:(BOOL)isEnabled
{
	[_textField setEnabled:isEnabled];
	[_stepper setEnabled:isEnabled];
	[_popUpButton setEnabled:isEnabled];
}

@end
//--------------------------------------------------------------------------------------------------------------------------------------------


//--------------------------------------------------------------------------------------------------------------------------------------------
@implementation THFileSizeStepperView : THValueAndUnitView

- (NSDictionary*)valueStepperInfos
{
	return @{@"maxValue":@(1000*1000)};
}

- (NSMenu*)unitMenu
{
	NSMenu *menu=[[NSMenu alloc] init];
	[menu addItem:[NSMenuItem th_menuItemWithTitle:THLocalizedString(@"FileSizePoids_KB") tag:1]];
	[menu addItem:[NSMenuItem th_menuItemWithTitle:THLocalizedString(@"FileSizePoids_MB") tag:2]];
	[menu addItem:[NSMenuItem th_menuItemWithTitle:THLocalizedString(@"FileSizePoids_GB") tag:3]];
	[menu addItem:[NSMenuItem th_menuItemWithTitle:THLocalizedString(@"FileSizePoids_TB") tag:4]];
	return menu;
}

@end

@implementation THDateWithinStepperView : THValueAndUnitView

- (NSDictionary*)valueStepperInfos
{
	return @{@"maxValue":@(1000)};
}

- (NSMenu*)unitMenu
{
	NSMenu *menu=[[NSMenu alloc] init];
	[menu addItem:[NSMenuItem th_menuItemWithTitle:THLocalizedString(@"Minute(s)") tag:THFileSearchUnit_min]];
	[menu addItem:[NSMenuItem th_menuItemWithTitle:THLocalizedString(@"Hour(s)") tag:THFileSearchUnit_hour]];
	[menu addItem:[NSMenuItem separatorItem]];
	[menu addItem:[NSMenuItem th_menuItemWithTitle:THLocalizedString(@"Day(s)") tag:THFileSearchUnit_day]];
	[menu addItem:[NSMenuItem th_menuItemWithTitle:THLocalizedString(@"Yeek(s)") tag:THFileSearchUnit_week]];
	[menu addItem:[NSMenuItem th_menuItemWithTitle:THLocalizedString(@"Month(s)") tag:THFileSearchUnit_month]];
	[menu addItem:[NSMenuItem th_menuItemWithTitle:THLocalizedString(@"Year(s)") tag:THFileSearchUnit_year]];
	return menu;
}

@end
//--------------------------------------------------------------------------------------------------------------------------------------------

File: Classes/THRuleEditorView/THValueAndUnitView.swift
// THValueAndUnitView.swift

import Cocoa

//--------------------------------------------------------------------------------------------------------------------------------------------
fileprivate class THValueAndUnitView_PopUpButton : NSPopUpButton {
	override var canBecomeKeyView: Bool { get { true } }
}

fileprivate class THValueAndUnitView_Stepper : NSStepper {
	override var canBecomeKeyView: Bool { get { true } }
}
//--------------------------------------------------------------------------------------------------------------------------------------------


//--------------------------------------------------------------------------------------------------------------------------------------------
@objc protocol THValueAndUnitViewDelegateProtocol {
	@objc func valueAndUnitViewDidChange(_ sender: THValueAndUnitView)
}

@objc class THValueAndUnitView : NSView, NSTextFieldDelegate {
	private var textField: NSTextField!
	private var stepper: NSStepper!
	private var popUpButton: NSPopUpButton!

	weak var controlView: THValueAndUnitViewDelegateProtocol!
	
	@objc init(frame: NSRect, popBezelStyle: NSButton.BezelStyle, controlSize: NSControl.ControlSize, controlView: THValueAndUnitViewDelegateProtocol) {
		super.init(frame: frame)

		self.controlView = controlView

		// faire height en control controlSize != small

		let fieldRect = NSRect(0.0, CGFloatFloor((frame.size.height - 19.0) / 2.0), 50.0, 19.0)
		let stepperRect = NSRect(fieldRect.size.width + 4.0 ,CGFloatFloor((frame.size.height - 22.0) / 2.0), 16.0, 22.0)
		var menuRect = NSRect(stepperRect.origin.x + stepperRect.size.width + 4.0, CGFloatFloor((frame.size.height - 22.0) / 2.0), 60.0, 22.0)

		if self is THDateWithinStepperView {
			menuRect.size.width = 90.0
		}

		let font = NSFont.systemFont(ofSize: NSFont.systemFontSize(for: controlSize))

		if true {
			textField = NSTextField(frame: fieldRect)

			textField.font = font
			//	result.target=self
			//	result.action=@selector(criterionTextFieldAction:)
			textField.delegate = self
			textField.isEditable = true
			textField.isSelectable = true
			textField.drawsBackground = true

			let nbFormatter = NumberFormatter()
			nbFormatter.numberStyle = .decimal
			nbFormatter.usesGroupingSeparator = false
			textField.formatter = nbFormatter

			let cell = textField.cell as! NSTextFieldCell
			cell.controlSize = controlSize
			cell.alignment = .right

			addSubview(textField)
		}

		if true {
			let stepInfos = self.valueStepperInfos()!

			stepper = THValueAndUnitView_Stepper(frame: stepperRect)
			stepper.target = self
			stepper.action = #selector(fileSizeStepperAction)
			stepper.minValue = 1.0
			stepper.integerValue = 1
			stepper.maxValue = Double(stepInfos["maxValue"] as! Int)
			
			let cell = stepper.cell as! NSStepperCell
			cell.controlSize = controlSize

			addSubview(stepper)
		}

		if true {
			let  menu  = self.unitMenu()!

			popUpButton = THValueAndUnitView_PopUpButton(frame: menuRect)
			popUpButton.alignment = .left
			popUpButton.bezelStyle = popBezelStyle
			popUpButton.target = self
			popUpButton.action = #selector(fileSizePoidsMenuAction)
			popUpButton.menu = menu
			popUpButton.autoresizingMask = [.maxXMargin]

			let cell = popUpButton.cell as! NSPopUpButtonCell
			cell.controlSize = controlSize
			cell.font = font

			addSubview(popUpButton)
		}
	}
	
	required init?(coder: NSCoder) {
		fatalError("init(coder:) has not been implemented")
	}
	
	deinit {
		controlView = nil
	}

	// MARK: -

	func valueStepperInfos() -> [String: Any]? {
		return nil
	}

	func unitMenu() -> NSMenu? {
		return nil
	}

	// MARK: -

	@objc func value() -> Int {
		return textField.integerValue
	}

	@objc func poids() -> Int {
		return popUpButton.selectedTag()
	}

	@objc func setValue(_ value: Int, poids: Int) {
		textField.integerValue = value
		stepper.integerValue = value
		popUpButton.selectItem(withTag: poids)
	}

	@objc func valueComps() -> String {
		return String(self.value()) + "|" + String(self.poids())
	}

	@objc func setValueComps(_ comps: String) {
		let c = comps.components(separatedBy: "|")
		if c.count == 2 {
			setValue(Int(c.first!)!, poids: Int(c.last!)!)
			return
		}
		setValue(1, poids: 1)
	}
	
	@objc func controlTextDidChange(_ notification: Notification) {
		let sender = notification.object as? NSTextField

		if sender == textField {
			if textField.integerValue >= Int(stepper.minValue) {
				stepper.integerValue = textField.integerValue
				controlView.valueAndUnitViewDidChange(self)
			}
		}
	}

	@objc func fileSizeStepperAction(_ sender: NSStepper) {
		textField.integerValue = sender.integerValue
		controlView.valueAndUnitViewDidChange(self)
	}

	@objc func fileSizePoidsMenuAction(_ sender: NSPopUpButton) {
		controlView.valueAndUnitViewDidChange(self)
	}

	@objc func setEnabled(_ enabled: Bool) {
		textField.isEnabled = enabled
		stepper.isEnabled = enabled
		popUpButton.isEnabled = enabled
	}

}
//--------------------------------------------------------------------------------------------------------------------------------------------


//--------------------------------------------------------------------------------------------------------------------------------------------
@objc class THFileSizeStepperView : THValueAndUnitView {
	
	override func  valueStepperInfos() -> [String: Any] {
		return ["maxValue": 1000 * 1000]
	}

	override func unitMenu() -> NSMenu {
		let menu = NSMenu()
		menu.addItem(NSMenuItem(withTitle: THLocalizedString("FileSizePoids_KB"), tag: 1))
		menu.addItem(NSMenuItem(withTitle: THLocalizedString("FileSizePoids_MB"), tag: 2))
		menu.addItem(NSMenuItem(withTitle: THLocalizedString("FileSizePoids_GB"), tag: 3))
		menu.addItem(NSMenuItem(withTitle: THLocalizedString("FileSizePoids_TB"), tag: 4))
		return menu
	}

}
//--------------------------------------------------------------------------------------------------------------------------------------------


//--------------------------------------------------------------------------------------------------------------------------------------------
@objc class THDateWithinStepperView : THValueAndUnitView {

	override func  valueStepperInfos() -> [String: Any] {
		return ["maxValue": 1000]
	}

	override func unitMenu() -> NSMenu {
		let menu = NSMenu()
		menu.addItem(NSMenuItem(withTitle: THLocalizedString("Minute(s)"), tag: THFileSearchUnit_min))
		menu.addItem(NSMenuItem(withTitle: THLocalizedString("Hour(s)"), tag: THFileSearchUnit_hour))
		menu.addItem(NSMenuItem.separator())
		menu.addItem(NSMenuItem(withTitle: THLocalizedString("Day(s)"), tag: THFileSearchUnit_day))
		menu.addItem(NSMenuItem(withTitle: THLocalizedString("Yeek(s)"), tag: THFileSearchUnit_week))
		menu.addItem(NSMenuItem(withTitle: THLocalizedString("Month(s)"), tag: THFileSearchUnit_month))
		menu.addItem(NSMenuItem(withTitle: THLocalizedString("Year(s)"), tag: THFileSearchUnit_year))
		return menu
	}

}
//--------------------------------------------------------------------------------------------------------------------------------------------
